psytran/family.py

```diff
@@ -161,3 +182,533 @@ def has_ancestor(node, node_type=Loop, inclusive=False, name=None):
+    # parallel section.
+    # Under the same parent node is causing a little confusion.
+    # The paths allow us to check index references for closeness
+    # Given a certain circumstance we can iron out the confusion
+    # Needs to be wrapped in a try else it fails
+    try:
+        path_to_omp_par = node_target.path_from(omp_ancestor_par)
+    except ValueError:
+        path_to_omp_par = False
+    try:
+        path_to_omp_par_do = node_target.path_from(omp_ancestor_par_do)
+    except ValueError:
+        path_to_omp_par_do = False
+
+    # DEFAULT result
+    # The default presence of OMP given the below ancestry
+    # True : There is an OpenMP Parallel section above the node,
+    #        This can be either a Parallel or Parallel Do section above
+    #        The default intended use is if this is True, do not do an
+    #        OpenMP transformation.
+    # False : There is no OpenMP Parallel sections detected above.
+    if ((omp_ancestor_par and path_to_omp_par)
+            or (omp_ancestor_par_do and path_to_omp_par_do)):
+        print("Parallel OMP region ancestor found")
+        omp_ancestry_presence = True
+    else:
+        omp_ancestry_presence = False
+
+    # When adding a OpenMP DO transformation, it's a little more nuanced.
+    # We will in principle reverse the result.
+    # True : There is an OpenMP Parallel Do section above the node.
+    # False : There is a Parallel section above the Node, and it is safe.
+    if transformation.omp_directive == "do":
+        # A OpenMP do transformation needs a parallel section above,
+        # it checks and corrects.
+        # An occurrence where adjacent parallel do regions are being
+        # picked up and the parallel ancestor is misreporting. Therefore,
+        # if there a parallel ancestor and path, but there is not a path to a
+        # parallel do reference node. Checking the paths mitigates this
+        # occurrence. To be reported to STFC as a bug in psyclone.
+        if omp_ancestor_par and path_to_omp_par and not path_to_omp_par_do:
+            print("Adjacent node detected in Ancestry")
+            print("Parallel OMP region ignored as transformation is OMP do")
+            omp_ancestry_presence = False
+
+        # If there is no path to both an parallel or parallel do Node,
+        # but it has reported an parallel ancestor
+        if not omp_ancestor_par and not path_to_omp_par:
+            print("No Parallel region present, cannot try do")
+            omp_ancestry_presence = True
+
+        # This stops the nesting of OMP do under parallel sections
+        # If there is one already present, it will effectively understand the
+        # above a parallel and a do, which we don't want to try parallelism in
+        # this occurrence.
+        if omp_ancestor_do:
+            omp_ancestry_presence = True
+
+    return omp_ancestry_presence
+
+
+def get_last_child_shed(loop_node):
```

Comment:
Any chance we could expand the name to `get_last_child_schedule`? Just because it's not immediately obvious that `shed` is meant to refer to `schedule`.